selfdrive/ui/qt/onroad.cc

```diff
@@ -511,24 +517,54 @@ void AnnotatedCameraWidget::drawLaneLines(QPainter &painter, const UIState *s) {
   }
 
   // paint path
-  QLinearGradient bg(0, height(), 0, height() / 4);
-  float start_hue, end_hue;
+//  double t = millis_since_boot();
+  QLinearGradient bg(0, height(), 0, 0);
   if (sm["controlsState"].getControlsState().getExperimentalMode()) {
-    const auto &acceleration = sm["modelV2"].getModelV2().getAcceleration();
-    float acceleration_future = 0;
-    if (acceleration.getZ().size() > 16) {
-      acceleration_future = acceleration.getX()[16];  // 2.5 seconds
-    }
-    start_hue = 60;
-    // speed up: 120, slow down: 0
-    end_hue = fmax(fmin(start_hue + acceleration_future * 45, 148), 0);
 
-    // FIXME: painter.drawPolygon can be slow if hue is not rounded
-    end_hue = int(end_hue * 100 + 0.5) / 100;
+    int track_vertices_len = scene.track_vertices.length();
+    assert(track_vertices_len % 2 == 0);
+    QVector<QPointF> right_points = scene.track_vertices.mid(0, track_vertices_len / 2);
+
+    const float MAX_UI_POINTS = 8;
+
+    // Filter out points out of frame
+    QVector<QPointF> valid_points;
+    std::copy_if(right_points.begin(), right_points.end(), std::back_inserter(valid_points),
+                 [&](const QPointF &point) { return point.y() >= 0 && point.y() <= height(); });
+    qDebug() << "valid_points len:" << valid_points.length() << "right_points len:" << right_points.length();
+
+    int drawn = 0;
+    int final_idx = 0;
+    if (!valid_points.isEmpty()) {
+      for (int p = 0; p < MAX_UI_POINTS; p++) {
+        int i = (p / (MAX_UI_POINTS - 1)) * (valid_points.length() - 1);
+        final_idx = i;
```

Comment:
this now ensures we: 1. always draw the last visible point, 2. always draws max 8 points, but think we can simplify this a bit and sacrifice those two by not enforcing them.

plus, when the path is very close to the vehicle, there will still be many points that are pretty much the same color, so we should enforce point distances instead of indexes.

Reply:
too complicated